src/input/src/joystick.rs

```diff
@@ -0,0 +1,115 @@
+
+//! Back-end agnostic joystick events.
+
+use std::any::Any;
+
+use { GenericEvent, JOYSTICK_AXIS };
+
+/// Components of a joystick button event. Not guaranteed consistent across
+/// backends.
+#[derive(Copy, Clone, RustcDecodable, RustcEncodable, PartialEq, Eq, Debug, Hash)]
+pub struct JoystickButton {
+    /// Which joystick was the button on.
+    pub id: i32,
+    /// Which button was pressed.
+    pub button: u8,
+
+}
+
+impl JoystickButton {
+    /// Create a new JoystickButton object. Intended for use by backends when
+    /// emitting events.
+    pub fn new(id: i32, button: u8) -> Self {
+        JoystickButton {
+            id: id,
+            button: button,
+        }
+    }
+}
+
+/// Components of a joystick axis move event. Not guaranteed consistent across
+/// backends.
+#[derive(Copy, Clone, RustcDecodable, RustcEncodable, PartialEq, Debug)]
+pub struct JoystickAxisArgs {
+    /// Which joystick moved.
+    pub id: i32,
+    /// The axis that moved.
+    pub axis: u8,
+    /// Position of the joystick. Usually [-1.0, 1.0], though backends may use
+    /// a different range for various devices.
+    pub position: f64
+}
+
+impl JoystickAxisArgs {
+    /// Create a new JoystickAxisArgs object. Intended for use by backends when
+    /// emitting events.
+    pub fn new(id: i32, axis: u8, position: f64) -> Self {
+        JoystickAxisArgs {
+            id: id,
+            axis: axis,
+            position: position,
+        }
+    }
+}
+
+/// The position of a joystick axis changed.
+pub trait JoystickAxisEvent: Sized {
+    /// Creates a joystick axis event.
+    fn from_id_axis_position(id: i32, axis: u8, position: f64, old_event: &Self) -> Option<Self>;
```

Comment:
It is used to transform one event into another of same type in generic code. You need the old event in case it uses a custom structure that contains extra information related to the event.

For example, the AI behavior tree uses this to change the update time interval https://github.com/PistonDevelopers/ai_behavior/blob/master/src/state.rs#L134